ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -914,12 +934,72 @@ static bool ShouldHandleFileOpEvent(
     {
         PerfSample findRootSample(perfTracer, PrjFSPerfCounter_FileOp_ShouldHandle_FindVirtualizationRoot);
         
-        *root = VirtualizationRoot_FindForVnode(
-            perfTracer,
-            PrjFSPerfCounter_FileOp_FindRoot,
-            PrjFSPerfCounter_FileOp_FindRoot_Iteration,
-            vnode,
-            context);
+        if (isDirectory)
+        {
+            if (KAUTH_FILEOP_RENAME == action)
+            {
+                // Directory renames into (or out) of virtualization roots require invalidating all of the entries
+                // within the directory being renamed (because all of those entires now have a new virtualzation root).
+                // Rather than trying to find all vnodes in the cache that are children of the directory being renamed
+                // we simply invalidate the entire cache.
+                //
+                // Potential future optimizations include:
+                //   - Only invalidate the cache if the rename moves a directory in or out of a directory
+                //   - Keeping a per-root generation ID in the cache to allow invalidating a subset of the cache
+                VnodeCache_InvalidateCache(perfTracer);
+            }
+            
+            *root = VnodeCache_FindRootForVnode(
+                perfTracer,
+                PrjFSPerfCounter_FileOp_Vnode_Cache_Hit,
+                PrjFSPerfCounter_FileOp_Vnode_Cache_Miss,
+                PrjFSPerfCounter_FileOp_FindRoot,
+                PrjFSPerfCounter_FileOp_FindRoot_Iteration,
+                vnode,
+                context);
+        }
+        else
+        {
+            // TODO(Mac): Once all hardlink paths are delivered to the appropriate root(s)
+            // check if the KAUTH_FILEOP_LINK case can be removed.  For now the check is required to make
+            // sure we're looking up the most up-to-date parent information for the vnode on the nexe
+            // access to the vnode cache
+            switch(action)
+            {
+                case KAUTH_FILEOP_LINK:
+                    *root = VnodeCache_InvalidateVnodeAndGetLatestRoot(
+                        perfTracer,
+                        PrjFSPerfCounter_FileOp_Vnode_Cache_Hit,
+                        PrjFSPerfCounter_FileOp_Vnode_Cache_Miss,
+                        PrjFSPerfCounter_FileOp_FindRoot,
+                        PrjFSPerfCounter_FileOp_FindRoot_Iteration,
+                        vnode,
+                        context);
+                    break;
+
+                case KAUTH_FILEOP_RENAME:
+                    *root = VnodeCache_RefreshRootForVnode(
```

Comment:
Remind me one last time the difference between invalidatevnodeAndGetLatestRoot and RefreshRootForVnode

Reply:
I'm renaming `VnodeCache_InvalidateVnodeAndGetLatestRoot` to `VnodeCache_InvalidateVnodeRootAndGetLatestRoot`.

The difference between the two:

- `VnodeCache_InvalidateVnodeAndGetLatestRoot` -> Invalidates the vnode's root in the cache (i.e. sets it to Interderminate)
- `VnodeCache_RefreshRootForVnode` -> Refreshes the vnode's root in the cache (i.e. looks up the root from disk and put that new root into the cache)